examples/static-transformer/Main.hs

```diff
@@ -62,53 +65,64 @@ import qualified Torch.TensorFactories         as D
 -- Transformer Language Model (GPT-2)
 --------------------------------------------------------------------------------
 
-data Activation (dtype :: D.DType)
+data Activation (dtype :: D.DType) (device :: (D.DeviceType, Nat))
  where
   Activation
-    :: forall dtype
-     . { unActivation :: forall shape . Tensor dtype shape -> Tensor dtype shape }
-    -> Activation dtype
+    :: forall dtype device
+     . { unActivation :: forall shape . Tensor device dtype shape -> Tensor device dtype shape }
+    -> Activation dtype device
 
-instance Show (Activation dtype) where
+instance Show (Activation dtype device) where
   show _ = mempty
 
-instance A.Parameterized (Activation dtype) where
+instance A.Parameterized (Activation dtype device) where
   flattenParameters _ = []
   replaceOwnParameters = return
 
-data MultiheadAttentionSpec (dtype :: D.DType)
-                            (embedDim :: Nat) (numHeads :: Nat)
+data MultiheadAttentionSpec (embedDim :: Nat) (numHeads :: Nat)
+                            (dtype :: D.DType)
+                            (device :: (D.DeviceType, Nat))
  where
   MultiheadAttentionSpec
     :: { mhDropoutProbSpec :: Double }
-    -> MultiheadAttentionSpec dtype embedDim numHeads
+    -> MultiheadAttentionSpec embedDim numHeads dtype device
  deriving (Show, Eq)
 
-data MultiheadAttention (dtype :: D.DType)
-                        (embedDim :: Nat) (numHeads :: Nat)
+data MultiheadAttention (embedDim :: Nat) (numHeads :: Nat)
+                        (dtype :: D.DType)
+                        (device :: (D.DeviceType, Nat))
  where
   MultiheadAttention
-    :: { mhInProj :: Linear dtype embedDim (embedDim * 3)
-       , mhOutProj :: Linear dtype embedDim embedDim
+    :: { mhInProj  :: Linear embedDim (embedDim * 3) dtype device
+       , mhOutProj :: Linear embedDim embedDim       dtype device
        , mhDropout :: Dropout
        }
-    -> MultiheadAttention dtype embedDim numHeads
+    -> MultiheadAttention embedDim numHeads dtype device
  deriving (Show, Generic)
 
 multiheadAttention
-  :: forall dtype embedDim numHeads seqLen batchSize headDim
+  :: forall embedDim numHeads seqLen batchSize headDim dtype device
    . ( 1 <= numHeads
      , embedDim ~ (headDim * numHeads)
      , Mod (embedDim * 3) 3 ~ 0
      , Div (embedDim * 3) 3 ~ embedDim
-     , KnownDType dtype
      , All KnownNat [embedDim, numHeads, seqLen, batchSize, headDim]
+     , KnownDType dtype
+     , dtype ~ DTypePromotion dtype (SumDType dtype)
+     , StandardFloatingPointDTypeValidation device dtype
+     , MatMulDTypeIsValid device dtype
+     , BasicArithmeticDTypeIsValid device dtype
+     , BasicArithmeticDTypeIsValid device (SumDType dtype)
```

Comment:
Where is BasicArithmeticDTypeIsValid 's definition?

Reply:
there's a new module called `Torch.Typed.Aux` where I put most of these definitions.
This particular one is defined in `Torch.Typed.Tensor`, though: https://github.com/tscholak/hasktorch/blob/feature/type-safe-device/hasktorch/src/Torch/Typed/Tensor.hs#L421-L426

Reply:
thx!